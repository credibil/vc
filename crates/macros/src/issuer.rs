use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::parse::{Error, Result};

use crate::parse::Data;

pub fn create_offer_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let subject_id = fields.remove("subject_id").unwrap();
    let credential_configuration_ids = fields.remove("credential_configuration_ids").unwrap();
    let pre_authorize = fields.remove("pre-authorize").unwrap();
    let tx_code_required = fields.remove("tx_code_required").unwrap();
    let send_type = fields.remove("send_type").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        CreateOfferRequest {
            credential_issuer: #credential_issuer.to_string(),
            subject_id: Some(#subject_id.to_string()),
            credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),
            pre_authorize: #pre_authorize,
            tx_code_required: #tx_code_required,
            send_type:  #send_type,
        }
    })
}

pub fn authorization_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let response_type = fields.remove("response_type").unwrap();
    let client_id = fields.remove("client_id").unwrap();
    let redirect_uri = fields.remove("redirect_uri").unwrap();
    let state = fields.remove("state").unwrap();
    //  let code_challenge = fields.remove("code_challenge").unwrap();
    let code_challenge_method = fields.remove("code_challenge_method").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        AuthorizationRequest {
            credential_issuer: #credential_issuer.to_string(),
            response_type: #response_type.to_string(),
            client_id: #client_id.to_string(),
            redirect_uri: Some(#redirect_uri.to_string()),
            state: Some(#state.to_string()),
            // code_challenge: #code_challenge.to_string(),
            code_challenge_method: #code_challenge_method.to_string(),

            // credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),

            ..Default::default()
        }
    })
}
