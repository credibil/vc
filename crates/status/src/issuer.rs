//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
