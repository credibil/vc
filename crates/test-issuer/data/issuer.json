{
    "credential_issuer": "http://credibil.io",
    "credential_endpoint": "http://credibil.io/credential",
    "deferred_credential_endpoint": "http://credibil.io/deferred",
    "display": {
        "name": "Credibil",
        "locale": "en-NZ"
    },
    "credential_configurations_supported": {
        "EmployeeID_JWT": {
            "format": "jwt_vc_json",
            "scope": "EmployeeIDCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "email"
                    ],
                    "display": [
                        {
                            "name": "Email",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address"
                    ],
                    "display": [
                        {
                            "name": "Residence",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "street_address"
                    ],
                    "display": [
                        {
                            "name": "Street Address",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "locality"
                    ],
                    "display": [
                        {
                            "name": "Locality",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "region"
                    ],
                    "display": [
                        {
                            "name": "Region",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "country"
                    ],
                    "display": [
                        {
                            "name": "Country",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "Developer_JWT": {
            "format": "jwt_vc_json",
            "scope": "DeveloperCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Developer",
                    "description": "Credibil certified developer credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/developer.png",
                        "alt_text": "Developer Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#010100",
                    "background_image": {
                        "uri": "http://credibil.io/assets/developer-background.png",
                        "alt_text": "Developer Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "DeveloperCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "proficiency"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Proficiency",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "org.iso.18013.5.1.mDL": {
            "format": "mso_mdoc",
            "doctype": "org.iso.18013.5.1.mDL",
            "scope": "DriverLicence",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "claims": [
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "birth_date"
                    ],
                    "mandatory": true
                },
                {
                    "path": [
                        "org.iso.18013.5.1.aamva",
                        "organ_donor"
                    ]
                }
            ]
        }
    }
}