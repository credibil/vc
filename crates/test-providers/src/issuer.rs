use anyhow::Result;
use credibil_core::blockstore::BlockStore;
use credibil_identity::{Identity, IdentityResolver, Key, SignerExt};
use credibil_se::{Algorithm, Signer};

use crate::blockstore::Mockstore;
use crate::identity::DidIdentity;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone)]
pub struct Issuer {
    identity: DidIdentity,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub async fn new(owner: &str) -> Self {
        Self {
            identity: DidIdentity::new(owner).await,
            blockstore: Mockstore::new(),
        }
    }
}

impl IdentityResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Identity> {
        self.identity.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.identity.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.identity.verifying_key().await
    }

    async fn algorithm(&self) -> Result<Algorithm> {
        Ok(self.identity.algorithm())
    }
}

impl SignerExt for Issuer {
    async fn verification_method(&self) -> Result<Key> {
        self.identity.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}
