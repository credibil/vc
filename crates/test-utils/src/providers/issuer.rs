use std::collections::HashMap;

use anyhow::anyhow;
use openid4vc::issuance::Issuer;
use provider::Result;

#[derive(Default, Clone, Debug)]
pub struct Store {
    issuers: HashMap<String, Issuer>,
}

impl Store {
    pub fn new() -> Self {
        let issuer = Issuer::sample();

        Self {
            issuers: HashMap::from([
                ("http://localhost:8080".into(), issuer.clone()),
                (issuer.credential_issuer.clone(), issuer),
            ]),
        }
    }

    pub fn get(&self, issuer_id: &str) -> Result<Issuer> {
        let Some(issuer) = self.issuers.get(issuer_id) else {
            return Err(anyhow!("issuer not found"));
        };
        Ok(issuer.clone())
    }
}
