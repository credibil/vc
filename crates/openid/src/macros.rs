
mod create_offer;

/// Construct a `serde_json::Value` from a JSON literal.
///
/// ```
/// # use serde_json::json;
/// #
/// let value = json!({
///     "code": 200,
///     "success": true,
///     "payload": {
///         "features": [
///             "serde",
///             "json"
///         ],
///         "homepage": null
///     }
/// });
/// ```
///
/// Variables or expressions can be interpolated into the JSON literal. Any type
/// interpolated into an array element or object value must implement Serde's
/// `Serialize` trait, while any type interpolated into a object key must
/// implement `Into<String>`. If the `Serialize` implementation of the
/// interpolated type decides to fail, or if the interpolated type contains a
/// map with non-string keys, the `json!` macro will panic.
///
/// ```
/// # use serde_json::json;
/// #
/// let code = 200;
/// let features = vec!["serde", "json"];
///
/// let value = json!({
///     "code": code,
///     "success": code == 200,
///     "payload": {
///         features[0]: features[1]
///     }
/// });
/// ```
///
/// Trailing commas are allowed inside both arrays and objects.
///
/// ```
/// # use serde_json::json;
/// #
/// let value = json!(["notice", "the", "trailing", "comma -->",]);
/// ```
#[macro_export]
macro_rules! create_offer_request {
    ({ $($json:tt)+ }) => {{
        let mut offer = $crate::issuer::CreateOfferRequest::default(); //serde_json::Map::new();
        $crate::internal_create_offer!(@object offer () ($($json)+) ($($json)+));
        offer
    }};
}

#[cfg(test)]
mod tests {
    use crate::issuer::SendType;

    const CREDENTIAL_ISSUER: &str = "http://vercre.io";

    #[test]
    fn create_offer() {
        let x = create_offer_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "credential_configuration_ids":["EmployeeID_JWT"],
            "subject_id": "normal_user",
            "pre-authorize": true,
            "tx_code_required": false,
            "send_type": SendType::ByRef,
            // "object": {
            //     "key": "value"
            // }
        });

        println!("{:?}", x);
    }
}