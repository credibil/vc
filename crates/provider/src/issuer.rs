#![allow(unused)]

use anyhow::Result;
use credibil_did::{DidResolver, Document};
use credibil_infosec::{Algorithm, Signer};
use credibil_vc::BlockStore;
use credibil_vc::status::issuer::Status;

use crate::blockstore::Mockstore;
use crate::keystore::Keyring;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone, Debug)]
pub struct Issuer {
    keyring: Keyring,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub fn new() -> Self {
        Self {
            keyring: Keyring::new(),
            blockstore: Mockstore::new(),
        }
    }
}

impl DidResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        self.keyring.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.keyring.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.keyring.verifying_key().await
    }

    fn algorithm(&self) -> Algorithm {
        self.keyring.algorithm()
    }

    async fn verification_method(&self) -> Result<String> {
        self.keyring.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

impl Status for Issuer {}
