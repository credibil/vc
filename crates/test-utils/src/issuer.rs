use chrono::{DateTime, Utc};
use serde::de::DeserializeOwned;
use serde::Serialize;
use vercre_datasec::{Algorithm, Decryptor, Encryptor, SecOps, Signer};
use vercre_did::{DidResolver, Document};
use vercre_openid::issuer::{
    Dataset, Client, Issuer, Metadata, Result, Server, StateStore, Subject,
};

use crate::store::keystore::IssuerKeystore;
use crate::store::{issuance, resolver, state};

pub const CREDENTIAL_ISSUER: &str = "http://vercre.io";
pub const CLIENT_ID: &str = "96bfb9cb-0513-7d64-5532-bed74c48f9ab";
pub const NORMAL_USER: &str = "normal_user";
pub const PENDING_USER: &str = "pending_user";

#[derive(Default, Clone, Debug)]
pub struct Provider {
    pub client: issuance::ClientStore,
    pub issuer: issuance::IssuerStore,
    pub server: issuance::ServerStore,
    pub subject: issuance::DatasetStore,
    pub state: state::Store,
}

impl Provider {
    #[must_use]
    pub fn new() -> Self {
        Self {
            client: issuance::ClientStore::new(),
            issuer: issuance::IssuerStore::new(),
            server: issuance::ServerStore::new(),
            subject: issuance::DatasetStore::new(),
            state: state::Store::new(),
        }
    }
}

impl vercre_openid::issuer::Provider for Provider {}

impl Metadata for Provider {
    async fn client(&self, client_id: &str) -> Result<Client> {
        self.client.get(client_id)
    }

    async fn register(&self, client: &Client) -> Result<Client> {
        self.client.add(client)
    }

    async fn issuer(&self, issuer_id: &str) -> Result<Issuer> {
        self.issuer.get(issuer_id)
    }

    async fn server(&self, server_id: &str) -> Result<Server> {
        self.server.get(server_id)
    }
}

impl Subject for Provider {
    /// Authorize issuance of the specified credential for the holder.
    async fn authorize(
        &self, subject_id: &str, credential_configuration_id: &str,
    ) -> Result<Vec<String>> {
        self.subject.authorize(subject_id, credential_configuration_id)
    }

    async fn dataset(&self, subject_id: &str, credential_identifier: &str) -> Result<Dataset> {
        self.subject.dataset(subject_id, credential_identifier)
    }
}

impl StateStore for Provider {
    async fn put(&self, key: &str, state: impl Serialize, dt: DateTime<Utc>) -> Result<()> {
        self.state.put(key, state, dt)
    }

    async fn get<T: DeserializeOwned>(&self, key: &str) -> Result<T> {
        self.state.get(key)
    }

    async fn purge(&self, key: &str) -> Result<()> {
        self.state.purge(key)
    }
}

impl DidResolver for Provider {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        resolver::resolve_did(url).await
    }
}

struct IssuerSec(IssuerKeystore);

impl SecOps for Provider {
    fn signer(&self, _identifier: &str) -> anyhow::Result<impl Signer> {
        Ok(IssuerSec(IssuerKeystore {}))
    }

    fn encryptor(&self, _identifier: &str) -> anyhow::Result<impl Encryptor> {
        Ok(IssuerSec(IssuerKeystore {}))
    }

    fn decryptor(&self, _identifier: &str) -> anyhow::Result<impl Decryptor> {
        Ok(IssuerSec(IssuerKeystore {}))
    }
}

impl Signer for IssuerSec {
    fn algorithm(&self) -> Algorithm {
        self.0.algorithm()
    }

    fn verification_method(&self) -> String {
        self.0.verification_method()
    }

    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.0.try_sign(msg)
    }
}

impl Encryptor for IssuerSec {
    async fn encrypt(&self, _plaintext: &[u8], _recipient_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }

    fn public_key(&self) -> Vec<u8> {
        todo!()
    }
}

impl Decryptor for IssuerSec {
    async fn decrypt(&self, _ciphertext: &[u8], _sender_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }
}
