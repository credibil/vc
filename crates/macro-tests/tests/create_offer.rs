#![allow(missing_docs)]

use insta::assert_yaml_snapshot as assert_snapshot;
use credibil_vc::issuer::SendType;
use macros::create_offer_request;

const CREDENTIAL_ISSUER: &str = "http://vercre.io";

#[test]
fn pre_authorized() {
    let subject_id = "normal_user";

    let request = create_offer_request!({
        "credential_issuer": CREDENTIAL_ISSUER,
        "credential_configuration_ids": ["EmployeeID_JWT"],
        "subject_id": &subject_id,
        "grant_types": ["urn:ietf:params:oauth:grant-type:pre-authorized_code"],
        "tx_code_required": true,
        "send_type": SendType::ByVal,
    });

    assert_snapshot!("pre-authorized", &request);
}

#[test]
fn default() {
    let request = create_offer_request!({
        "credential_issuer": "http://vercre.io",
        "credential_configuration_ids": ["EmployeeID_JWT"],
    });

    assert_snapshot!("default", &request);
}
