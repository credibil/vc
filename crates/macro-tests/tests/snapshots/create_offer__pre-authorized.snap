---
source: crates/macro-tests/tests/create_offer.rs
assertion_line: 22
expression: "&request"
---
credential_issuer: "http://credibil.io"
subject_id: normal_user
credential_configuration_ids:
  - EmployeeID_JWT
grant_types:
  - "urn:ietf:params:oauth:grant-type:pre-authorized_code"
tx_code_required: true
send_type: ByVal
